architecture/evm/error_normalizer.go

```diff
@@ -276,29 +324,29 @@ func ExtractJsonRpcError(r *http.Response, nr *common.NormalizedResponse, jr *co
 						nil,
 						details,
 					),
+					true, // retryable
 				)
 			}
-		} else if code == -32602 ||
+		}
+
+		//----------------------------------------------------------------
+		// 11) "Invalid Argument / Params" errors
+		//----------------------------------------------------------------
+
+		if code == -32602 ||
 			strings.Contains(msg, "param is required") ||
 			strings.Contains(msg, "Invalid Request") ||
 			strings.Contains(msg, "validation errors") ||
 			strings.Contains(msg, "invalid argument") ||
-			strings.Contains(msg, "invalid params") {
-			if strings.Contains(msg, "tx of type") {
-				// TODO For now we're returning a server-side exception for this error
-				// so that the request is retried on a different upstream which supports the requested TX type.
-				// This must be properly handled when "Upstream Features" is implemented which allows for feature-based routing.
-				return common.NewErrEndpointServerSideException(
-					common.NewErrJsonRpcExceptionInternal(
-						int(code),
-						common.JsonRpcErrorCallException,
-						err.Message,
-						nil,
-						details,
-					),
-					nil,
-				)
-			}
+			strings.Contains(msg, "invalid params") ||
+			strings.Contains(msg, "tx of type") ||
+			strings.Contains(msg, "invalid type") {
+
+			// by default, we retry this type of client-side exception, as the root cause
+			// might be an invalid argument, that another upstream might support.
+			//
+			// Examples:
+			// - Envio not supporting { blockNumber: XXX, blockHash: YYY } for eth_getBlockReceipts
```

Comment:
For majority of the cases this is probably not correct. Validation errors or invalid params can happen a lot if user provides a bad HEX number for example. I suggest to treat these vendor-specific ones explicitly vs retrying all 4xx errors.

A counter-argument is we don't expect properly designed prod dApps to generate these kinds of errors as they'll have higher level validation checks before calling RPC. Although it's in theory, best is to check big RPC clients how many 4xx do they see.

What do you think?

Reply:
totally agreed. I think the invalid arg, params, request is too broad to always retry. would be nice though to get as many of these invalid `arg -> still retryable` error messages as possible.